### DISCUSSION TITLE
FIPS provider checksum will not work on iOS devices

### DISCUSSION THREAD

**Discussion by danbey** [openssl/openssl#25200] [Q&A] [answered]
Hi All,

I want to share my research conclusion that fips support will not work on openssl running on ios device.

- To understand the background, iOS has a procedure of signing every file they are adding to the application bundle. This signing must be done with a certified developerID. This is all part of the build process of the app.

- Each time a file is signed it get a different fingerprint (different checksum)

- So, if fips.so will be part of the app it will have to be singed and each time we'll get a new fingerprint result. This fingerprint will, of course, override the last fingerprint that fips.so had.

- On the other hand, in openssl as part of building openssl, we creates a fingerprint/checksum for the fips.so and putting the fingerprint/checksum (under the field "module-mac") in the openssl.cnf .

- So, we conclude that the openssl.cnf module-mac will not be identical to the fips.so fingerprint because the iOS signing process changed it.

To see this in action, this is the debug prints from the openssl when trying to load fips.so.
Please notice the "out" field which is the checksum that openssl gets from the fips.so vs the "expected" that openssl gets from the "module-mac" field in the openssl.cnf

This is for the first build:
DEBUG  *** out:
84 ec d0 49 68 3e 60 b4 a3 f0 fe 14 00 01 9f 11 9e c9 15 b3 66 c0 c3 97 a1 a2 7e 1a c4 fb d2 4b 
DEBUG  *** expected:
ee 1c 9a 4a ef bb fd 75 ae 35 1f 02 5e 8f 34 01 9f 1f 56 18 dc 58 b3 ca 98 3b f1 b9 88 47 7d 04 

This is for the second build:
DEBUG *** out:
3d 22 d2 29 01 4f 00 bf 1c 9f 58 8c 94 5e 14 1b b2 b6 76 78 65 33 9d 2b 8f 48 f4 5a 94 d9 d1 4f 
DEBUG *** expected:
ee 1c 9a 4a ef bb fd 75 ae 35 1f 02 5e 8f 34 01 9f 1f 56 18 dc 58 b3 ca 98 3b f1 b9 88 47 7d 04 

This is for the third build:
DEBUG *** out:
a5 45 71 9c d9 43 fe 13 87 f0 30 cc 18 57 c7 db c8 14 f7 51 7f d2 69 8b 02 f7 93 04 81 2a 8e 03 
DEBUG *** expected:
ee 1c 9a 4a ef bb fd 75 ae 35 1f 02 5e 8f 34 01 9f 1f 56 18 dc 58 b3 ca 98 3b f1 b9 88 47 7d 04 

 
To sum up, IMHO making security checks on iOS devices is not needed because it's already made by the iOS build process.

Thanks,
Dan

**Comment by t8m** [comment, marked as answer]
Yes, that is a known issue. I do not know if the signature validation done by the iOS itself could be acceptable as an integrity check for the FIPS 140 purposes. If so, it would be possible to simply remove the internal integrity check of the fips provider for iOS builds. Otherwise the module MAC in the fipsmodule.cnf would have to be updated after the fips provider signing.

 The fips provider is not validated on iOS and without changing the security policy and the implementation it cannot be.

**Reply by mattcaswell** [reply]
> I do not know if the signature validation done by the iOS itself could be acceptable as an integrity check for the FIPS 140 purposes.

Doesn't the integrity check have to be done within the module boundary using FIPS validated crypto?

**Reply by danbey** [reply]
Hi, Thanks for the response!

1. I know that the signing procedure can only made with a certifide developerID. JonDo can't change/replace the fips.dylib once build and signing was past (again, using a certifide DeveloperID) and the app bundle was created. So why do we need to do the verified again within opnessl context ? But i prefere a senior iOS develoer could verifiy this .
2. IMHO I think removing the possibility of building openssl for ios with --fips-enabled is something to consider.
3. Maybe this issue should be discuss in a more wide forum. I'll be happy to attend to contribute my experience on this issue.

Thanks,
Dan

**Reply by t8m** [reply]
> Doesn't the integrity check have to be done within the module boundary using FIPS validated crypto?

It definitely needs to be done using FIPS validated crypto. Whether one module can do a check of another module - I do not know, a lab would have to be consulted for that.

**Reply by t8m** [reply]
Actually we did something - at least partially - similar in RHEL. When there still was an OpenSSH module, it called OpenSSL for the integrity check which was a separate FIPS module. But that was a long time ago and rules could have changed since.

**Reply by paulidale** [reply]
My guess would be that a second module could do the check but the result would have to be transported over a secure channel which is more work than moving the check saves.

This is a question for your FIPS lab.  They are the only ones who can provide a canonical answer as to what's to be done.
